sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceTestBase.scala

```diff
@@ -457,3 +458,340 @@ case class SessionUpdate(
+  def getKeyToNumValuesSchemas(): (StructType, StructType) = {
+    val metadata = getKeyToNumValuesSchemasWithMetadata()
+    (metadata.keySchema, metadata.valueSchema)
+  }
+
+  /**
+   * @return ColumnFamilyMetadata with schema and KeyEncoderSpec
+   */
+  def getKeyToNumValuesSchemasWithMetadata(): ColumnFamilyMetadata = {
+    val keySchema = StructType(Array(
+      StructField("key", IntegerType)
+    ))
+    val valueSchema = StructType(Array(
+      StructField("value", LongType)
+    ))
+    val encoderSpec = NoPrefixKeyStateEncoderSpec(keySchema)
+    ColumnFamilyMetadata(keySchema, valueSchema, encoderSpec)
+  }
+
+  /**
+   * @param stateVersion The join state format version:
+   * @return A tuple of (keySchema, valueSchema)
+   */
+  def getKeyWithIndexToValueSchemas(stateVersion: Int): (StructType, StructType) = {
+    val metadata = getKeyWithIndexToValueSchemasWithMetadata(stateVersion)
+    (metadata.keySchema, metadata.valueSchema)
+  }
+
+  /**
+   * @param stateVersion The state format version
+   * @return ColumnFamilyMetadata with schema and KeyEncoderSpec
+   */
+  def getKeyWithIndexToValueSchemasWithMetadata(stateVersion: Int): ColumnFamilyMetadata = {
+    val keySchema = StructType(Array(
+      StructField("key", IntegerType, nullable = false),
+      StructField("index", LongType)
+    ))
+
+    val valueSchema = if (stateVersion == 2 || stateVersion == 3) {
+      StructType(Array(
+        StructField("value", IntegerType, nullable = false),
+        StructField("time", TimestampType, nullable = false),
+        StructField("matched", BooleanType)
+      ))
+    } else {
+      StructType(Array(
+        StructField("value", IntegerType, nullable = false),
+        StructField("time", TimestampType, nullable = false)
+      ))
+    }
+
+    val encoderSpec = NoPrefixKeyStateEncoderSpec(keySchema)
+    ColumnFamilyMetadata(keySchema, valueSchema, encoderSpec)
+  }
+
+  /**
+   * Returns all schemas for stream-stream join V3 (multi-column family) in legacy 2-tuple format.
+   * V3 uses a single state store with multiple column families instead of separate stores.
+   *
+   * @return Map of column family name to (keySchema, valueSchema)
+   */
+  def getJoinV3ColumnSchemaMap(): Map[String, (StructType, StructType)] = {
+    getJoinV3ColumnSchemaMapWithMetadata().view.mapValues { metadata =>
+      (metadata.keySchema, metadata.valueSchema)
+    }.toMap
+  }
+
+  /**
+   * @return Map of column family name to ColumnFamilyMetadata
+   */
+  def getJoinV3ColumnSchemaMapWithMetadata(): Map[String, ColumnFamilyMetadata] = {
+    val (keyToNumKeySchema, keyToNumValueSchema) = getKeyToNumValuesSchemas()
+    val (keyWithIndexKeySchema, keyWithIndexValueSchema) = getKeyWithIndexToValueSchemas(3)
+
+    val keyToNumEncoderSpec = NoPrefixKeyStateEncoderSpec(keyToNumKeySchema)
+    val keyWithIndexEncoderSpec = NoPrefixKeyStateEncoderSpec(keyWithIndexKeySchema)
+
+    Map(
+      KEY_TO_NUM_VALUES_LEFT -> ColumnFamilyMetadata(
```

Comment:
can use `getStateStoreName(Left, KeyToNumValuesType)` in join manager. Same for others